File: 19.js
class DoubleHashSet {
  static EMPTY = 0;
  static OCC = 1;
  static DEL = 2;

  constructor(m) {
    this.m = m;
    this.keys = new Array(m).fill(null);
    this.state = new Array(m).fill(DoubleHashSet.EMPTY);
    this.distinct = 0;
  }

  h1(k) {
    let v = k % this.m;
    if (v < 0) v += this.m;
    return v;
  }

  h2(k) {
    return 1 + (Math.abs(k) % (this.m - 1));
  }

  search(k) {
    const a = this.h1(k), b = this.h2(k);
    for (let i = 0; i < this.m; i++) {
      const idx = (a + i * b) % this.m;
      if (this.state[idx] === DoubleHashSet.EMPTY) return false;
      if (this.state[idx] === DoubleHashSet.OCC && this.keys[idx] === k) return true;
    }
    return false;
  }

  insert(k) {
    if (this.search(k)) return;

    const a = this.h1(k), b = this.h2(k);
    let firstDel = -1;

    for (let i = 0; i < this.m; i++) {
      const idx = (a + i * b) % this.m;

      if (this.state[idx] === DoubleHashSet.EMPTY) {
        const pos = firstDel !== -1 ? firstDel : idx;
        this.keys[pos] = k;
        this.state[pos] = DoubleHashSet.OCC;
        this.distinct++;
        return;
      }
      if (this.state[idx] === DoubleHashSet.DEL && firstDel === -1) {
        firstDel = idx;
      }
    }

    if (firstDel !== -1) {
      this.keys[firstDel] = k;
      this.state[firstDel] = DoubleHashSet.OCC;
      this.distinct++;
    }
  }
}

// Driver
const fs = require("fs");
const data = fs.readFileSync(0, "utf8").trim().split(/\s+/);
let idx = 0;

const m = parseInt(data[idx++]);
const q = parseInt(data[idx++]);

const s = new DoubleHashSet(m);

for (let i = 0; i < q; i++) {
  const x = parseInt(data[idx++]);
  s.insert(x);
  console.log(s.distinct);
}

File: 19.py
class DoubleHashSet:
    EMPTY = 0
    OCC = 1
    DEL = 2

    def __init__(self, m):
        self.m = m
        self.keys = [None] * m
        self.state = [DoubleHashSet.EMPTY] * m
        self.distinct = 0

    def _h1(self, k):
        return k % self.m

    def _h2(self, k):
        return 1 + (abs(k) % (self.m - 1))

    def search(self, k):
        a = self._h1(k)
        b = self._h2(k)
        for i in range(self.m):
            idx = (a + i * b) % self.m
            if self.state[idx] == DoubleHashSet.EMPTY:
                return False
            if self.state[idx] == DoubleHashSet.OCC and self.keys[idx] == k:
                return True
        return False

    def insert(self, k):
        if self.search(k):
            return  # already exists

        a = self._h1(k)
        b = self._h2(k)

        first_del = -1
        for i in range(self.m):
            idx = (a + i * b) % self.m
            if self.state[idx] == DoubleHashSet.EMPTY:
                insert_pos = first_del if first_del != -1 else idx
                self.keys[insert_pos] = k
                self.state[insert_pos] = DoubleHashSet.OCC
                self.distinct += 1
                return
            if self.state[idx] == DoubleHashSet.DEL and first_del == -1:
                first_del = idx

        # fallback (insert at first tombstone)
        if first_del != -1:
            self.keys[first_del] = k
            self.state[first_del] = DoubleHashSet.OCC
            self.distinct += 1


def main():
    import sys
    data = sys.stdin.read().strip().split()
    m = int(data[0])
    q = int(data[1])
    s = DoubleHashSet(m)

    idx = 2
    for _ in range(q):
        x = int(data[idx]); idx += 1
        s.insert(x)
        print(s.distinct)

if __name__ == "__main__":
    main()
